[Internal] Documentation: Adds Hedging Documentation (#4699)
Azure/azure-cosmos-dotnet-v3: docs/Cross Region Request Hedging.md

> # Cross Region Hedging APIs + Samples

Reviewer: How about align with the theme of the API which is availability strategy? 
I.e. E2E latency can be reliability achieved through cross region hedging and call out reads as scope.

> Cross Region Hedging Availability Strategies are a new feature in the Cosmos SDK enables the sending of redundant parallel requests to multiple regions during high latency periods. This feature aims to help lower latency and improve availability in your applications. The SDK will send the first request to the primary region. If there is no response from the backend before the threshold time, then the SDK will begin sending hedged requests to the regions in order of the `ApplicationPreferredRegions` list. After the first hedged request is sent out, the hedged requests will continue to be fired off one by one after waiting for the time specified in the threshold step. Once a response is received from one of the requests, the availability strategy will check to see if the result is considered final. If the result is final, then it is returned. If not, the SDK will skip the remaining threshold/threshold step time and send out the next hedged request. If all hedged requests are sent out and no final response is received, the SDK will return the last response it received. The AvaiabilityStrategy operates on the RequestInvokerHandler level meaning that each hedged request will go through its own [handler pipeline](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/docs/SdkDesign.md#handler-pipeline), including the ClientRetryPolicy. This means that the hedged requests will be retried independently of each other. Note that the hedged requests are restricted to the region they are sent out in so no cross region retries will be made, only local retries. The primary request however, will behave as a normal request.

Reviewer 2: ```suggestion
The ability to specify a Cross Region Hedging Availability Strategy is a new feature in the Cosmos SDK that enables the sending of redundant parallel requests to multiple regions during high latency periods. This feature aims to help lower latency and improve availability in your applications. The SDK will send the first request to the primary region. If there is no response from the backend before the threshold time, then the SDK will begin sending hedged requests to the regions in order of the `ApplicationPreferredRegions` list. After the first hedged request is sent out, the hedged requests will continue to be fired off one by one after waiting for the time specified in the threshold step. Once a response is received from one of the requests, the availability strategy will check to see if the result is considered final. If the result is final, then it is returned. If not, the SDK will skip the remaining threshold/threshold step time and send out the next hedged request. If all hedged requests are sent out and no final response is received, the SDK will return the last response it received. The `AvailabilityStrategy` operates on the `RequestInvokerHandler` level meaning that each hedged request will go through its own [handler pipeline](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/docs/SdkDesign.md#handler-pipeline), including the `ClientRetryPolicy`. This means that the hedged requests will be retried independently of each other. Note that the hedged requests are restricted to the region they are sent out in so no cross region retries will be made, only local retries. The primary request however, will behave as a normal request.
```

Adjusted the beginning to try to make declension match a bit better.

> Cross Region Hedging Availability Strategies are a new feature in the Cosmos SDK enables the sending of redundant parallel requests to multiple regions during high latency periods. This feature aims to help lower latency and improve availability in your applications. The SDK will send the first request to the primary region. If there is no response from the backend before the threshold time, then the SDK will begin sending hedged requests to the regions in order of the `ApplicationPreferredRegions` list. After the first hedged request is sent out, the hedged requests will continue to be fired off one by one after waiting for the time specified in the threshold step. Once a response is received from one of the requests, the availability strategy will check to see if the result is considered final. If the result is final, then it is returned. If not, the SDK will skip the remaining threshold/threshold step time and send out the next hedged request. If all hedged requests are sent out and no final response is received, the SDK will return the last response it received. The AvaiabilityStrategy operates on the RequestInvokerHandler level meaning that each hedged request will go through its own [handler pipeline](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/docs/SdkDesign.md#handler-pipeline), including the ClientRetryPolicy. This means that the hedged requests will be retried independently of each other. Note that the hedged requests are restricted to the region they are sent out in so no cross region retries will be made, only local retries. The primary request however, will behave as a normal request.

Reviewer: > The SDK will send the first request to the primary region
Primary might get mis-interpreted, how about preferred first region.

> Cross Region Hedging Availability Strategies are a new feature in the Cosmos SDK enables the sending of redundant parallel requests to multiple regions during high latency periods. This feature aims to help lower latency and improve availability in your applications. The SDK will send the first request to the primary region. If there is no response from the backend before the threshold time, then the SDK will begin sending hedged requests to the regions in order of the `ApplicationPreferredRegions` list. After the first hedged request is sent out, the hedged requests will continue to be fired off one by one after waiting for the time specified in the threshold step. Once a response is received from one of the requests, the availability strategy will check to see if the result is considered final. If the result is final, then it is returned. If not, the SDK will skip the remaining threshold/threshold step time and send out the next hedged request. If all hedged requests are sent out and no final response is received, the SDK will return the last response it received. The AvaiabilityStrategy operates on the RequestInvokerHandler level meaning that each hedged request will go through its own [handler pipeline](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/docs/SdkDesign.md#handler-pipeline), including the ClientRetryPolicy. This means that the hedged requests will be retried independently of each other. Note that the hedged requests are restricted to the region they are sent out in so no cross region retries will be made, only local retries. The primary request however, will behave as a normal request.

Reviewer: Thought: How about represent the pattern through mermaid like flow instead?

Author: Is the mermaid flow at the bottom of the document ok?

Reviewer: Flow below is more design, idea here is can we replace above text with a logical mermaid flow. 
Its not a blocker.

Reviewer: Can be a follow up, merging this PR

> The example below will create a `CosmosClient` instance with AvailabilityStrategy enabled with a 500ms threshold. This means that if a request takes longer than 500ms the SDK will send a new request to the backend in order of the Preferred Regions List. If the `ApplicationRegion` or `ApplicationPreferredRegions` list is not set, then an AvailabilityStrategy will not be able to applied. If still no response comes back from the first hedge or the primary request after the step time, another parallel request will be made to the next region.  The SDK will then return the first *final* response that comes back from the backend. The threshold parameter is a required parameter can be set to any value greater than 0. There is also an option to the `AvailabilityStrategy` at request level and override the client level `AvailabilityStrategy` by setting the `AvailabilityStrategy` on the `RequestOptions` object.

Reviewer 3: >If still no response comes back from the first hedge or the primary request after the step time, another parallel request will be made to the next region.

maybe 'successive requests are started to each of the other regions at intervals defined by X, for as long as no final response has yet been obtained'

> The example below will create a `CosmosClient` instance with AvailabilityStrategy enabled with a 500ms threshold. This means that if a request takes longer than 500ms the SDK will send a new request to the backend in order of the Preferred Regions List. If the `ApplicationRegion` or `ApplicationPreferredRegions` list is not set, then an AvailabilityStrategy will not be able to applied. Parallel requests to the remaining regions will be sent at 100ms intervals defined by the `thresholdStep` parameter until a final response is found or all regions are exausted. The SDK will then return the first *final* response that comes back from the backend, if there are no final resposnes, the SDK will return the last result it recieved. The `threshold` parameter is a required parameter can be set to any value greater than 0. There is also an option to add the `AvailabilityStrategy` at request level, overriding the client level `AvailabilityStrategy`, by setting an `AvailabilityStrategy` in the `RequestOptions` object.

Reviewer 2: ```suggestion
The example below will create a `CosmosClient` instance with `AvailabilityStrategy` enabled with a 500ms threshold. This means that if a request takes longer than 500ms the SDK will send a new request to the backend in order of the Preferred Regions List. If the `ApplicationRegion` or `ApplicationPreferredRegions` list is not set, then an `AvailabilityStrategy` will not be able to applied. Parallel requests to the remaining regions will be sent at 100ms intervals defined by the `thresholdStep` parameter until a final response is found or all regions are exhausted. The SDK will then return the first *final* response that comes back from the backend, if there are no final responses, the SDK will return the last result it received. The `threshold` parameter is a required parameter and can be set to any value greater than 0. There is also an option to add the `AvailabilityStrategy` at request level, overriding the client level `AvailabilityStrategy`, by setting an `AvailabilityStrategy` in the `RequestOptions` object.
```

> The example below will create a `CosmosClient` instance with AvailabilityStrategy enabled with a 500ms threshold. This means that if a request takes longer than 500ms the SDK will send a new request to the backend in order of the Preferred Regions List. If the `ApplicationRegion` or `ApplicationPreferredRegions` list is not set, then an AvailabilityStrategy will not be able to applied. Parallel requests to the remaining regions will be sent at 100ms intervals defined by the `thresholdStep` parameter until a final response is found or all regions are exausted. The SDK will then return the first *final* response that comes back from the backend, if there are no final resposnes, the SDK will return the last result it recieved. The `threshold` parameter is a required parameter can be set to any value greater than 0. There is also an option to add the `AvailabilityStrategy` at request level, overriding the client level `AvailabilityStrategy`, by setting an `AvailabilityStrategy` in the `RequestOptions` object.

Reviewer: Values listed in our docs will be literally used, 100ms is too aggressive. 
How about changing default to at-least order of seconds

## APIs

### Enable `AvailabilityStrategy` at client level 

When Building a new `CosmosClient` there will be an option to include a Cross Region Hedging Availability Strategy in that client.

```csharp
CosmosClient client = new CosmosClientBuilder("connection string")
    .WithApplicationPreferredRegions(

>         new List<string> { "East US", "Central US", "West US" } )

Reviewer 3: ```suggestion
        new { "East US", "Central US", "West US" } )
```
or collection syntax

    .WithAvailabilityStrategy(
        AvailabilityStrategy.CrossRegionHedgingStrategy(
        threshold: TimeSpan.FromMilliseconds(500),
        thresholdStep: TimeSpan.FromMilliseconds(100)
     ))
    .Build();
```

or

```csharp
CosmosClientOptions options = new CosmosClientOptions()
{
    AvailabilityStrategy
     = AvailabilityStrategy.CrossRegionHedgingStrategy(
        threshold: TimeSpan.FromMilliseconds(500),
        thresholdStep: TimeSpan.FromMilliseconds(100)

>      )

Reviewer 2: ```suggestion
     ),
```

      ApplicationPreferredRegions = new List<string>() { "East US", "West US", "Central US"},
};

CosmosClient client = new CosmosClient(
    accountEndpoint: "account endpoint",
    authKeyOrResourceToken: "auth key or resource token",
    clientOptions: options);
```

> > Note: `ApplicationRegion` or `ApplicationPreferredRegions` MUST be set to add an `AvailabilityStrategy`.

Reviewer 2: What happens if not?  An exception from the constructor?

Reviewer: My preference is to still leave documentation as required and be explicit.

### Override client level `AvailabilityStrategy` or add `AvailabilityStrategy` at request level:

```csharp
//Send one request out with a more aggressive threshold
ItemRequestOptions requestOptions = new ItemRequestOptions()
{

>     AvailabilityStrategyOptions =AvailabilityStrategy.CrossRegionHedgingStrategy(

Reviewer 3: ```suggestion
    AvailabilityStrategyOptions = AvailabilityStrategy.CrossRegionHedgingStrategy(
```

        threshold: TimeSpan.FromMilliseconds(100),
        thresholdStep: TimeSpan.FromMilliseconds(50)
     ))
};
```

#### Disable client level `AvailabilityStrategy`:

```csharp
//Send one request out without an AvailabilityStrategy
ItemRequestOptions requestOptions = new ItemRequestOptions()
{
    AvailabilityStrategyOptions =AvailabilityStrategy.DisabledStrategy()
};
```

When enabled at the `CosmosClient` level, the availability strategy applies to all read requests unless explicitly disabled per request: ReadItem, Queries (single and cross partition), ReadMany, and ChangeFeed. It is not enabled for write requests.

## Diagnostics

> In the diagnostics data there are two new areas of note `Response Region` and `Hedge Context` that will appear when using this feature. `Response Region` shows the region that the request is ultimately served out of. `Hedge Context` shows all the regions requests were sent to.

Reviewer: Please use a sample references

Author: Will add a sample

# Reliability achieve low end to end latency for read request with cross region hedging

> The ability to specify a Cross Region Hedging Availability Strategy is a new feature in the Cosmos SDK that enables the sending of redundant parallel requests to multiple regions during high latency periods. This feature aims to help lower latency and improve availability in your applications. The SDK will send the first request to the primary region. If there is no response from the backend before the threshold time, then the SDK will begin sending hedged requests to the regions in order of the `ApplicationPreferredRegions` list. After the first hedged request is sent out, the hedged requests will continue to be fired off one by one after waiting for the time specified in the threshold step. Once a response is received from one of the requests, the availability strategy will check to see if the result is considered final. If the result is final, then it is returned. If not, the SDK will skip the remaining threshold/threshold step time and send out the next hedged request. If all hedged requests are sent out and no final response is received, the SDK will return the last response it received. The `AvailabilityStrategy` operates on the `RequestInvokerHandler` level meaning that each hedged request will go through its own [handler pipeline](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/docs/SdkDesign.md#handler-pipeline), including the `ClientRetryPolicy`. This means that the hedged requests will be retried independently of each other. Note that the hedged requests are restricted to the region they are sent out in so no cross region retries will be made, only local retries. The primary request however, will behave as a normal request.

Reviewer 4: Let's mention the tradeoffs a bit- RU consumption is multiplied by number of hedged requests sent out. I borrowed this language from the [Java SDK doc](https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/performance-tips-java-sdk-v4?tabs=api-async#threshold-based-availability-strategy). Also split up the paragraph a bit

```suggestion
The ability to specify a Cross Region Hedging Availability Strategy is a new feature in the Cosmos SDK that enables the sending of redundant parallel requests to multiple regions during high latency periods. This feature can lower latency and improve availability in scenarios where a particular region is slow or temporarily unavailable, but it may incur more cost in terms of request units when parallel cross-region requests are required. 

When the cross region hedging strategy is enabled, the SDK will send the first request to the primary region. If there is no response from the backend before the threshold time, then the SDK will begin sending hedged requests to the regions in order of the `ApplicationPreferredRegions` list. After the first hedged request is sent out, the hedged requests will continue to be fired off one by one after waiting for the time specified in the threshold step. Once a response is received from one of the requests, the availability strategy will check to see if the result is considered final. If the result is final, then it is returned. If not, the SDK will skip the remaining threshold/threshold step time and send out the next hedged request. If all hedged requests are sent out and no final response is received, the SDK will return the last response it received. 

The `AvailabilityStrategy` operates on the `RequestInvokerHandler` level meaning that each hedged request will go through its own [handler pipeline](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/docs/SdkDesign.md#handler-pipeline), including the `ClientRetryPolicy`. This means that the hedged requests will be retried independently of each other. Note that the hedged requests are restricted to the region they are sent out in so no cross region retries will be made, only local retries. The primary request however, will behave as a normal request.
```

> The example below will create a `CosmosClient` instance with `AvailabilityStrategy` enabled with a 1500ms threshold. This means that if a request takes longer than 1500ms the SDK will send a new request to the backend in order of the Preferred Regions List. If the `ApplicationRegion` or `ApplicationPreferredRegions` list is not set, then an `AvailabilityStrategy` will not be able to applied. Parallel requests to the remaining regions will be sent at 1000ms intervals defined by the `thresholdStep` parameter until a final response is found or all regions are exhausted. The SDK will then return the first *final* response that comes back from the backend, if there are no final responses, the SDK will return the last result it received. The `threshold` parameter is a required parameter and can be set to any value greater than 0. There is also an option to add the `AvailabilityStrategy` at request level, overriding the client level `AvailabilityStrategy`, by setting an `AvailabilityStrategy` in the `RequestOptions` object.

Reviewer 4: This description should go under the "Enable `Availability Strategy` at client level" heading

## APIs

### Enable `AvailabilityStrategy` at client level

When Building a new `CosmosClient` there will be an option to include a Cross Region Hedging Availability Strategy in that client.

```csharp
CosmosClient client = new CosmosClientBuilder("connection string")
    .WithApplicationPreferredRegions(
        new List<string> { "East US", "Central US", "West US" } )
    .WithAvailabilityStrategy(
        AvailabilityStrategy.CrossRegionHedgingStrategy(

>         threshold: TimeSpan.FromMilliseconds(1500),

Reviewer: ```suggestion
        threshold: TimeSpan.FromSeconds(1.5),
```

>         thresholdStep: TimeSpan.FromMilliseconds(1000)

Reviewer: ```suggestion
        thresholdStep: TimeSpan.FromSeconds(1000)
```
